src/compiler/checker.ts

```diff
@@ -28240,7 +28240,29 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                     matching || type,
                     checkDerived ?
                         t => isTypeDerivedFrom(t, c) ? t : isTypeDerivedFrom(c, t) ? c : neverType :
-                        t => isTypeStrictSubtypeOf(t, c) ? t : isTypeStrictSubtypeOf(c, t) ? c : isTypeSubtypeOf(t, c) ? t : isTypeSubtypeOf(c, t) ? c : neverType,
+                        t => {
+                            // WAS: isTypeStrictSubtypeOf(t, c) ? t : isTypeStrictSubtypeOf(c, t) ? c : isTypeSubtypeOf(t, c) ? t : isTypeSubtypeOf(c, t) ? c : neverType;
+
+                            // Avoid going through relations for comparison between literals of the same types;
+                            // if they're not equal then they're not related and we can skip doing a bunch of work.
+                            if (t.flags & TypeFlags.Literal && (t.flags & TypeFlags.Literal) === (c.flags & TypeFlags.Literal)) {
+                                if (t.flags & TypeFlags.BooleanLiteral) {
+                                    return t === c ? t : neverType;

```

Comment:
`true` and `false` have fresh and un-fresh versions (and since this is control flow, you can probably witness both, since that's the point of having two) - so you can't *quite* only check `===` equality - there's 2 true's and 2 false's.

```suggestion
                                    return ((t === trueType || t === regularTrueType) ? (c === trueType || c === regularTrueType) : (c === falseType || c === regularFalseType)) ? t : neverType;

```

Actually, it's somewhat surprising we don't have an `isSameBooleanLiteralValue` function lying around, but I guess it just doesn't come up too often?

Reply:
Ah, yes, I forgot about those. I should really pull this into a function.

Reply:
We probably don't need the helper, because the first thing isRelatedTo does is to strip freshness.

Reply:
Also... as long as I don't allow this code to work when two enum literals are being compared, this entire optimization works when brought into `isTypeRelatedTo`. I might commit that to run perf on it quick as that's kinda interesting. But, I suspect that it will actually hurt the case I'm trying to fix because it's still going to make 4 calls to relations even though it doesn't need to.

It's kinda like relations should say something about both the positive and negative case, e.g. answer stuff like "A and B are not related, and also B is not related to A so no need to call me later".